security/pkg/server/ca/server.go

```diff
@@ -182,36 +183,49 @@ func (s *Server) HandleCSR(ctx context.Context, request *pb.CsrRequest) (*pb.Csr
 
 // Run starts a GRPC server on the specified port.
 func (s *Server) Run() error {
-	listener, err := net.Listen("tcp", fmt.Sprintf(":%d", s.port))
-	if err != nil {
-		return fmt.Errorf("cannot listen on port %d (error: %v)", s.port, err)
-	}
+	grpcServer := s.grpcServer
+	var listener net.Listener
+	var err error
+
+	if grpcServer == nil {
+		listener, err = net.Listen("tcp", fmt.Sprintf(":%d", s.port))
+		if err != nil {
+			return fmt.Errorf("cannot listen on port %d (error: %v)", s.port, err)
+		}
 
-	var grpcOptions []grpc.ServerOption
-	grpcOptions = append(grpcOptions, s.createTLSServerOption(), grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor))
+		var grpcOptions []grpc.ServerOption
+		grpcOptions = append(grpcOptions, s.createTLSServerOption(), grpc.UnaryInterceptor(grpc_prometheus.UnaryServerInterceptor))
 
-	grpcServer := grpc.NewServer(grpcOptions...)
+		grpcServer = grpc.NewServer(grpcOptions...)
+	}
 	pb.RegisterIstioCAServiceServer(grpcServer, s)
 	pb.RegisterIstioCertificateServiceServer(grpcServer, s)
 
 	grpc_prometheus.EnableHandlingTimeHistogram()
 	grpc_prometheus.Register(grpcServer)
 
-	// grpcServer.Serve() is a blocking call, so run it in a goroutine.
-	go func() {
-		serverCaLog.Infof("Starting GRPC server on port %d", s.port)
+	if listener != nil {
+		// grpcServer.Serve() is a blocking call, so run it in a goroutine.
+		go func() {
+			serverCaLog.Infof("Starting GRPC server on port %d", s.port)
 
-		err := grpcServer.Serve(listener)
+			err := grpcServer.Serve(listener)
 
-		// grpcServer.Serve() always returns a non-nil error.
-		serverCaLog.Warnf("GRPC server returns an error: %v", err)
-	}()
+			// grpcServer.Serve() always returns a non-nil error.
+			serverCaLog.Warnf("GRPC server returns an error: %v", err)
+		}()
+	}
 
 	return nil
 }
-
 // New creates a new instance of `IstioCAServiceServer`.
 func New(ca CertificateAuthority, ttl time.Duration, forCA bool,
+		hostlist []string, port int, trustDomain string, sdsEnabled bool) (*Server, error) {
+			return NewIstiod(nil, ca, ttl, forCA, hostlist, port, trustDomain, sdsEnabled)
+}
+
+// New creates a new instance of `IstioCAServiceServer`, running inside an existing gRPC server.
```

Comment:
s/New/NewIstiod

Reply:
NewIstiod is not accurate here, this method creates a new ca server.